docs: initial managed vault specs (#52)
noble-assets/dollar: spec/vaults-v2/01_state_vaults_v2.md

> # Vaults V2 State

Reviewer: ```suggestion
# Managed Vault State
```

## NAV (Net Asset Value)

> The `NAV` field is a [`collections.Item`][item] that stores the current Net Asset Value of the single Noble vault, representing the total value of all assets including its remote positions and inflight funds (`math.LegacyDec`).

Reviewer: ```suggestion
The `NAV` field is a [`collections.Item`][item] that stores the current Net Asset Value, representing the total value of all assets including its remote positions and inflight funds (`math.LegacyDec`).
```

The NAV calculation includes:

> - Local assets (held in the Noble vault)

Reviewer: ```suggestion
- Local assets (held in the vault)
```

- Remote position values (capital deployed from the vault)
- Inflight funds (in transit between the vault and remote positions)
- Minus pending liabilities (withdrawals to be paid)

```go
const NAVKey = []byte("vaults/v2/nav")
```

### NAV Calculation Formula

```
Total NAV = Local Assets
          + Σ(Remote Position Values)
          + Σ(Inflight Funds Values)
          - Pending Withdrawal Liabilities

NAV per Share = Total NAV / Total Outstanding Shares
```

## LastNAVUpdate

The `LastNAVUpdate` field is a [`collections.Item`][item] that stores the timestamp of the last NAV calculation (`time.Time`).

```go
const LastNAVUpdateKey = []byte("vaults/v2/last_nav_update")
```

## RemotePositions

> The `RemotePositions` field is a mapping ([`collections.Map`][map]) between position ID (`uint64`) and a `vaults.v2.RemotePosition` value. The single Noble vault maintains multiple remote positions across different protocols and chains.

Reviewer: ```suggestion
The `RemotePositions` field is a mapping ([`collections.Map`][map]) between position ID (`uint64`) and a `vaults.v2.RemotePosition` value. We maintain multiple remote positions across different protocols and chains.
```

```go
const RemotePositionPrefix = []byte("vaults/v2/remote_position/")
```

### RemotePosition Structure

```go
type RemotePosition struct {
    PositionID       uint64
    VaultAddress     []byte  // Address of the ERC-4626 compatible vault (Boring Vault or other)
    ChainID          uint32  // Hyperlane Domain ID (e.g., 998 for Hyperliquid, 8453 for Base, 4000261 for Noble App Layer)
    SharesHeld       math.Int // Number of remote vault shares held
    Principal        math.Int // USDN amount initially deposited
    LastUpdatedNAV   math.LegacyDec
    LastUpdateTime   time.Time
    Status           PositionStatus // ACTIVE, WITHDRAWING, CLOSED
}

> ```

Reviewer: Let's remove this until we have the Protobuf finalized!

```suggestion
TODO: Add structure definition once finalized
```

```go
const InflightFundsPrefix = []byte("vaults/v2/inflight_funds/")
```

### InflightFund Structure

```go
type InflightFund struct {
    HyperlaneRouteID uint32       // Hyperlane route identifier
    TransactionID    string       // Hyperlane message ID
    Type             InflightType // DEPOSIT_TO_POSITION, WITHDRAWAL_FROM_POSITION, REBALANCE_BETWEEN_POSITIONS, PENDING_DEPLOYMENT, PENDING_WITHDRAWAL_DISTRIBUTION
    Amount           math.Int     // Always in USDN
    SourceDomain     uint32       // Hyperlane source domain
    DestDomain       uint32       // Hyperlane destination domain
    SourcePosition   *uint64      // Optional: source position ID if from position
    DestPosition     *uint64      // Optional: destination position ID if to position
    InitiatedAt      time.Time
    ExpectedAt       time.Time
    Status           InflightStatus // PENDING, CONFIRMED, COMPLETED, FAILED
    ValueAtInitiation math.Int    // USDN value when initiated
}

> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

```go
const InflightValueByRoutePrefix = []byte("vaults/v2/inflight_value_by_route/")
```

## PendingDeploymentFunds

The `PendingDeploymentFunds` field is a [`collections.Item`][item] that stores the amount of USDN received from deposits but not yet deployed to remote positions (`math.Int`).

```go
const PendingDeploymentFundsKey = []byte("vaults/v2/pending_deployment")
```

## PendingWithdrawalDistribution

The `PendingWithdrawalDistribution` field is a [`collections.Item`][item] that stores the amount of USDN returned from remote positions but not yet distributed to withdrawal claimants (`math.Int`).

```go
const PendingWithdrawalDistributionKey = []byte("vaults/v2/pending_withdrawal_dist")
```

## WithdrawalQueue

The `WithdrawalQueue` field is a mapping ([`collections.Map`][map]) between a withdrawal request ID (`uint64`) and a `vaults.v2.WithdrawalRequest` value.

```go
const WithdrawalQueuePrefix = []byte("vaults/v2/withdrawal_queue/")
```

### WithdrawalRequest Structure

> ```go
> type WithdrawalRequest struct {
>     RequestID        uint64
>     User             []byte
>     SharesAmount     math.Int
>     RequestedAmount  math.Int  // Amount in $USDN requested
>     NAVAtRequest     math.LegacyDec
>     Timestamp        time.Time
>     Status           WithdrawalStatus // PENDING, PROCESSING, CLAIMABLE, CLAIMED
>     FulfilledAmount  math.Int  // Actual amount available for claim
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## WithdrawalQueueSequence

The `WithdrawalQueueSequence` field is a [`collections.Sequence`][sequence] that generates unique withdrawal request IDs.

```go
const WithdrawalQueueSequenceKey = []byte("vaults/v2/withdrawal_queue_seq")
```

## PendingWithdrawals

The `PendingWithdrawals` field is a [`collections.Item`][item] that stores the total amount of pending withdrawals for the vault (`math.Int`).

```go
const PendingWithdrawalsKey = []byte("vaults/v2/pending_withdrawals")
```

## UserShares

The `UserShares` field is a mapping ([`collections.Map`][map]) between user address (`[]byte`) and their share balance in the vault (`math.Int`).

```go
const UserSharesPrefix = []byte("vaults/v2/user_shares/")
```

## TotalShares

The `TotalShares` field is a [`collections.Item`][item] that stores the total outstanding shares for the vault (`math.Int`).

```go
const TotalSharesKey = []byte("vaults/v2/total_shares")
```

## DepositLimits

The `DepositLimits` field is a [`collections.Item`][item] that stores the deposit limits for the vault (`vaults.v2.DepositLimit`).

```go
const DepositLimitsKey = []byte("vaults/v2/deposit_limits")
```

### DepositLimit Structure

> ```go
> type DepositLimit struct {
>     MaxDepositPerUser     math.Int
>     MaxDepositPerBlock    math.Int
>     MaxTotalDeposits      math.Int
>     MinDepositAmount      math.Int
>     DepositCooldownBlocks uint64
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## UserDepositHistory

The `UserDepositHistory` field is a mapping ([`collections.Map`][map]) between a composite key of user address and block height (`[]byte`, `int64`) and deposit amount (`math.Int`). Used for malicious deposit detection.

```go
const UserDepositHistoryPrefix = []byte("vaults/v2/user_deposit_history/")
```

## DepositVelocity

The `DepositVelocity` field is a mapping ([`collections.Map`][map]) between user address (`[]byte`) and a `vaults.v2.DepositVelocity` value for tracking deposit patterns and defending against manipulation.

```go
const DepositVelocityPrefix = []byte("vaults/v2/deposit_velocity/")
```

### DepositVelocity Structure

> ```go
> type DepositVelocity struct {
>     LastDepositBlock      int64
>     RecentDepositCount    uint32  // Rolling count over time window
>     RecentDepositVolume   math.Int // Rolling volume over time window
>     SuspiciousActivityFlag bool
>     TimeWindowBlocks      int64   // e.g., 1000 blocks
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## RemotePositionOracles

The `RemotePositionOracles` field is a mapping ([`collections.Map`][map]) between a position ID (`string`) and oracle configuration (`vaults.v2.PositionOracleConfig`). Each remote position has a dedicated oracle that provides share price and shares held data.

```go
const RemotePositionOraclesPrefix = []byte("vaults/v2/remote_position_oracles/")
```

### PositionOracleConfig Structure

> ```go
> type PositionOracleConfig struct {
>     PositionID        string   // Unique identifier for the position
>     OracleAddress     string   // Oracle contract address on remote chain
>     ChainID           uint32   // Source chain ID
>     VaultAddress      string   // Address of vault holding this position
>     AuthorizedSender  string   // Authorized message sender address
>     MaxStaleness      int64    // Maximum blocks before data considered stale
>     LastUpdate        time.Time // Timestamp of last oracle update
>     SharePrice        sdk.Dec  // Current share price from oracle
>     SharesHeld        sdk.Int  // Number of shares held in position
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## OracleUpdates

The `OracleUpdates` field is a mapping ([`collections.Map`][map]) between a composite key of chain ID and vault address (`uint32`, `[]byte`) and the latest price update (`vaults.v2.OracleUpdate`). This stores the most recent share price data received directly from the Hyperlane Mailbox.

```go
const OracleUpdatesPrefix = []byte("vaults/v2/oracle_updates/")
```

### OracleUpdate Structure

> ```go
> type OracleUpdate struct {
>     SharePrice       math.LegacyDec  // Current price per share
>     TotalShares      math.Int        // Total shares outstanding
>     TotalAssets      math.Int        // Total assets in vault
>     Timestamp        time.Time       // When price was read on remote chain
>     BlockHeight      int64           // Noble block when Mailbox processed update
>     SourceDomain     uint32          // Hyperlane domain ID of source
>     MessageID        string          // Hyperlane message ID from Mailbox
>     Status           UpdateStatus    // VALIDATED, APPLIED (no PENDING - direct processing)
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## OracleRouteConfigs

The `OracleRouteConfigs` field is a mapping ([`collections.Map`][map]) between Hyperlane route ID (`uint64`) and route configuration (`vaults.v2.OracleRouteConfig`). Each route represents a specific path for oracle data transmission through the Hyperlane Mailbox.

```go
const OracleRouteConfigsPrefix = []byte("vaults/v2/oracle_route_configs/")
```

### OracleRouteConfig Structure

> ```go
> type OracleRouteConfig struct {
>     RouteID              uint64
>     SourceDomain         uint32   // Source Hyperlane domain
>     DestinationDomain    uint32   // Destination Hyperlane domain (Noble)
>     UpdateFrequency      uint32   // Expected blocks between updates
>     MaxStaleness         uint32   // Maximum blocks before stale
>     AuthorizedSender     []byte   // Authorized oracle contract on source chain
>     LastMessageNonce     uint32   // Last processed message nonce from Mailbox
>     LastProcessedTime    time.Time // When last Mailbox message was processed
>     Active               bool     // Whether route is currently active
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```

## LastOracleUpdate

The `LastOracleUpdate` field is a mapping ([`collections.Map`][map]) between position ID (`uint64`) and the timestamp of the last successful oracle update (`time.Time`). This is updated when the Mailbox handler processes a price update for the position.

```go
const LastOracleUpdatePrefix = []byte("vaults/v2/last_oracle_update/")
```

## VaultConfiguration

The `VaultConfiguration` field is a [`collections.Item`][item] that stores the configuration for the Noble vault (`vaults.v2.VaultConfig`).

```go
const VaultConfigurationKey = []byte("vaults/v2/vault_config")
```

### VaultConfig Structure

> ```go
> type VaultConfig struct {
>     MaxRemotePositions    uint32
>     AllowedChains         []uint32  // Hyperlane domain IDs
>     AllowedVaultAddresses map[uint32][][]byte // Chain ID -> List of approved vault addresses
>     RebalanceThreshold    math.LegacyDec // Percentage deviation before rebalance
>     WithdrawalDelayBlocks int64
>     ManagementFee         math.LegacyDec // Annual fee as percentage
>     PerformanceFee        math.LegacyDec // Fee on profits
>     EmergencyMode         bool
>     MaxInflightDuration   int64          // Max blocks funds can be inflight
>     InflightValueCap      math.Int       // Maximum value allowed inflight
> }
> ```

Reviewer: Ditto!

```suggestion
TODO: Add structure definition once finalized
```
